### ISSUE TITLE
Use slippy.Grid to support tiles in different projections systems

### ISSUE THREAD

**Issue by ear7h** [go-spatial/tegola#668]
This issue is for a design of using the new [`slippy.Grid`](https://godoc.org/github.com/go-spatial/geom/slippy#Grid) in tegola, in order for it to output multiple projection systems. Implementation of this functionality would close #243, #267, and #660.

Some issues to consider for this design are:
1. where should the output projection system be specified? (url query string, config, command line?)
2. where in the code should the conversion take place? and how should the info from (1) get there?
3. Depending on (1) how should the cache work?
    * the same `/:map/:layer/:z/:x/:y` could have multiple tiles with different projections

(I'll post my own later this week)

**Comment by ARolek** [comment]
I know @jj0hns0n has some opinions on this with some of the work he's been prototyping.

**Comment by jj0hns0n** [comment]
@ear7h lets have a quick chat in slack. Definitely have some opinions about 1 & 3 based on work for the OGC testbed.

**Comment by ear7h** [comment]
After talking to @jj0hns0n it looks like we should add support for [OGC API tiles](http://docs.opengeospatial.org/DRAFTS/19-zzz.html). It looks like:
```
{
    "description": "OGC API Tiles",
    "links": [
        {
            "href": "http://tiles.example.com/maps/{tileMatrixSetId}/{tileMatrix}/{tileCol}/{tileRow}.pbf",
            "rel": "item",
            "title": "Mapbox vector tiles",
            "type": "application/vnd.mapbox-vector-tile"
        }
    ],
    "tileMatrixSetLinks": [
        {
            "tileMatrixSet": "WorldCRS84Quad",
            "tileMatrixSetURI": "http://schemas.opengis.net/tms/1.0/json/examples/WorldCRS84Quad.json"
        },
        {
            "tileMatrixSet": "WebMercatorQuad",
            "tileMatrixSetURI": "http://schemas.opengis.net/tms/1.0/json/examples/WebMercatorQuad.json"
        }
    ],
    "title": "OGC-API-Tiles"
}
```

This format is helpful in understanding how one client might deal with multiple projection systems. One rule we that should guide the implementation is **each served tile set must have the same srid for all of its layers but, a map configured in tegola can result in multiple served tile sets**. An implementation might entail:
* In the config, a map would have a `projections` array parameter to take all the SRIDs
    - there would also be configurations like `srid_format` and `srid_query` to tell tegola how the
        the composite map name (name + srid) will be constructed or the name of the url query var
        respectively (the two options are mutually exclusive).
* an array of SRIDs would be part of each `atlas.Map`
    - a single `atlas.Map` might actually return multiple maps for serving as a tilejson/ogc api
    - the srid becomes part of the key in the `cache`
* Projection conversion would happen in `atlas.Map.EncodeMVTile` where there is currently a
    check for it

**Comment by tomass** [comment]
Wouldn't it be simpler if tegola would simply support ONE projection per installation (or one projection per map)? If I understand correctly, upon receiving a request of tile: "scale/x/y" you have to calculate extent of the tile and use it for postgis mvt call?

You could specify tileMatrix info mentioned above in this issue in configuration file (basically extent and scale:resolutions info, tilesize) and that would be enough to calculate tile extent?

**Comment by callsumzg** [comment]
We would need this feature in order to use Tegola in our environment. I am currently evaluating vector tile servers that must meet the following requirements:

- cache tiles
- pre-generate tiles
- PostGIS integration with custom projection (EPSG: 2056)
- output tiles in a custom projection (EPSG: 2056)

The latter is crucial for us, as we operate a WebGIS with official government data, all of which is published (and maintained) in LV95 (EPSG: 2056). Our WebGIS is based on OpenLayers, which can handle EPSG: 2056 without any issues. However, it cannot handle multiple projections simultaneously—at least, to my knowledge, vector tiles cannot be reprojected on-the-fly. And even if this were possible, it would be resource-intensive and come at the cost of performance.

There are already vector tile services that allow custom projections as output, e.g., TiPg. Unfortunately, it does not support caching or pre-generating tiles.

**Comment by iwpnd** [comment]
Tegola does not support custom projections and adding them is quite the effort. While it seems fun to work on, I don’t think that any of the current active folk in the community will find the time to add it.

**Comment by tomass** [comment]
I understand what @callsumzg is saying, at the time the only viable option (in open source) is to use OpenLayers ant they do not allow to reproject vector tiles on the fly. So the only option is to serve local projection vector tiles. This is a problem in Lithuania too (epsg:3346).

Now if we restrict scope to have source only in MVT_PostGIS, then all of the MVT part is done by PostGIS. The only thing which has to be done is:
* have a configuration of custom projection: a set of scales (resolutions) (I guess that should be enough, projection information should not be needed), something similar to WMTS configuration could be used.
* ability to specify custom projection in tile configuration (say in current setting of srid)
* use above mention info when calculating extent information on calling PostGIS ST_asMVT.

Would such a scope be acceptable to Tegola?

If so, would somebody be able to be a guide/mentor for such a work?

**Comment by ARolek** [comment]
@callsumzg I think this is technically possible, but you will have to do some work to make all the pieces dance. [Here's a quick example of of using the `mvt_postgis` provider with a 4326 projection](https://github.com/go-spatial/tegola/tree/master/mvtprovider/postgis#example-mvt_postgis-and-map-config-for-srid-4326). Note this is not exactly relevant, but demonstrates how some of the geoprocessing functions can be used to work towards your goal. 

A few things to note:

* If your data lives in a different projection than 3857 then you will need to reproject the incoming bbox to match your data's projection. For example in the SQL: `ST_Transform(!BBOX!, 2056)` or by calling `ST_TileEnvelope(:z, :x, :y, :srid)`.
* When your data is reprojected to tile coordinate space using `ST_AsMVTGeom` you will need to pass in the bounds made in the previous step.
* The internal tegola viewer will not work as it uses maplibre for rendering which is expecting tiles in 3857 projection by default. You will need to use another renderer and define the coordinate space on the client side so it can interpret the tiles correctly.

I might be missing something, but happy to help navigate this discussion. If we land on a solid solution I would request that we get some documentation together for the future community members that need to embark on this same journey. 

References:

* https://github.com/go-spatial/tegola/issues/731#issuecomment-756941064
* https://postgis.net/docs/ST_AsMVTGeom.html
* https://postgis.net/docs/ST_TileEnvelope.html

**Comment by ARolek** [comment]
@tomass your comment came in while I was writing mine. Take a look at my write up and see if it's helpful. I think we can accomplish this objective with the various pieces available now.

**Comment by iwpnd** [comment]
What is the reason you need the data in the origin projection? What is keeping you from reprojecting the origin geometry?

**Comment by iwpnd** [comment]
@callsumzg okay, for this purpose it makes sense. 
If TipG works for you, then seeding tiles, ist just requesting tiles + storing tiles in a cache. That’s fairly simple to implement. I want to take a look at TiPG now. 👀

**Comment by tomass** [comment]
@iwpnd if you want to use vector tiles from tegola in OL (the only one multi-projection open webmap framework), the only way is to have your (main) map projection as 3857.

**RASTER DISTORTION:** This means that ALL other data sources have to be reprojected to 3857 which would not be that much of a problem with today's resolutions, but 3857 scales (or zoom levels if you wish) can be very different from local ones which does cause visual bad results when "catching" closest scale (think of pre cached WMTS). For raster the problem is distortion.

**REQUIREMENT:** In other cases there is a fixed requirement to have a map in local projection AND specific (non x^n) scales (but rather something imagined as 1000, 2000, 5K, 10K, 20K, 50K etc).

**Comment by tomass** [comment]
> @tomass your comment came in while I was writing mine. Take a look at my write up and see if it's helpful. I think we can accomplish this objective with the various pieces available now.

@ARolek if I understand correctly, you make an assumption, that BBOX would return correct extent given z,x,y of a tile. But as far as I know that is not the case, because currently the calculation is for mercator. If local projection BBOX extent for zxy would be the same, then we would have no problems as coordinates in vector tiles are always [0; 4096].

In reality correct extent (bbox) calculation is the main part which would have to be developed. With BBOX calculated everything else would be easy :-) For bbox we would need some array of local projection resolutions. Something like you can find in say WMTS (something which is set up in openlayers tilegrid)
Say:
Level ID: 0 Resolution: 1322.9193125052918 (Scale: 5000000)
Level ID: 1 Resolution: 793.7515875031751 (Scale: 3000000)
Level ID: 2 Resolution: 529.1677250021168 (Scale: 2000000)
etc.

And then you would use resolution with zxy to calculate actual extent of a tile.

Years ago I've written some demo code to explain tile extent calculations, don't know if they are correct but results looked ok :-)
So given resolution array (for 4 zoom levels)
```
res float[] = ARRAY[2.6458386250105836,  -- 0
                    1.3229193125052918,  -- 1
                    0.5291677250021167,  -- 2
                    0.26458386250105836  -- 3
                   ];
```
Tile size, origin x and origin y:
```
c_tilesize int = 512;
c_origin_x float = 307558.9491752772;
c_origin_y float = 6264605.2051798785;
```

I wold then calculate tile geometry (so tile extent) for given 
```
l_tile_geom = st_setsrid(st_makepolygon(st_geomfromtext('LINESTRING(' ||
  (x       * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - (y       * res[z] * c_tilesize) || ',' ||
  ((x + 1) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - (y       * res[z] * c_tilesize) || ',' ||
  ((x + 1) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y + 1) * res[z] * c_tilesize) || ',' ||
  (x       * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y + 1) * res[z] * c_tilesize) || ',' ||
  (x       * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - (y       * res[z] * c_tilesize) ||
  ')')), 3346);
```
And this geometry is the BBOX for specific local projection.

Basically we have a cartesian space [x;y] in local projection, so we have some "starting" point for tiles as defined by origin x and origin y variables (this is used when for one reason or another an area around 0;0 is not actually used for data). We would then divide the space from origin point using zoom levels resolution and tile size.
![paveikslas](https://github.com/user-attachments/assets/0527b3bf-b663-4897-b319-887abe07cc45)


So configurations for all above numbers for projections would have to go somewhere (or they could probably come from projection info).

**Comment by tomass** [comment]
Wait... but we could simply change sql from:

`sql = "SELECT ST_AsMVTGeom(geom,!BBOX!) AS geom, gid FROM gis.landuse WHERE geom && !BBOX!"`

to:

`sql = "SELECT ST_AsMVTGeom(geom,calculate_my_projection_bbox(!z!,!x!,!y!)) AS geom, gid FROM gis.landuse WHERE geom && calculate_my_projection_bbox(!z!,!x!,!y!)"`

In that case the only thing we need is a database function calculate_my_projection_bbox(!z!,!x!,!y!)... and it should work.

**Comment by tomass** [comment]
Played a little bit around it and in principle it works (orange vector tile from tegola is depicted on a 3346 WMTS layer)
![paveikslas](https://github.com/user-attachments/assets/80cde72d-68ce-480e-b9c7-10c0e86fc070)

What have I done:
* in OpenLayers defined TileGrid in order for local WMTS to work as it does use local scales (I would have to do it anyway, independent of vector tiles)
* created a database function as per my code above with the resolutions as in TileGrid
* changed sql in tegola configuration to use this function instead of !BBOX! both when calling ST_AsMVTGeom as well as in WHERE clause.

What I've found/encountered
* !BBOX! is mandatory, so I've added a pointless "and !BBOX! is not null" in WHERE clause
* not all X/Y values are ok, tegola complains about Y of > 400 and Y ~15 - had to use _origin: [0,6500000]_ in OL tilegrid and then use the same origin in db function (this allows me to kind of control the XY values, but is not a good solution as it will probably not work in all required "places").

My questions:
1. Where does validation of XY come from and what is it trying to do?
2. Would it be a wanted addition to tegola? It looks like it would only require to extend functionality of !BBOX! calculation using origin point and an array of resolutions.

**Comment by ARolek** [comment]
@tomass thanks for sharing your research! This is great news as it sounds like most of the pieces are available to make this work. Let me answer some of your questions and then I have a few more for you.

> Where does validation of XY come from and what is it trying to do?

I believe this is in the `geom` package's `slippy` package under the method `Extent`. This package relies on the `proj` package which might need to be extended to account for various SRIDs.

https://github.com/go-spatial/geom/blob/55e666d67bc5969a5f2e09b17185f72727e5f0f9/slippy/tile_grid.go#L48

> Would it be a wanted addition to tegola? It looks like it would only require to extend functionality of !BBOX! calculation using origin point and an array of resolutions.

Ideally the `geom` package would be able to support different projections like you're exploring here. I'm not familiar enough with projection math to make these enhancements. Is this in your wheelhouse? If so, it would be a fantastic addition.

> !BBOX! is mandatory, so I've added a pointless "and !BBOX! is not null" in WHERE clause

Hmm, that's annoying. We should check for the use of `!BBOX!` or `!Z!` `!X!` `!Y!`

A couple of follow up questions:

Did you have to implement a custom `calculate_my_projection_bbox(!z!,!x!,!y!)` function? I figured you could accomplish this 2 ways:

```sql
ST_Transform(ST_TileEnvelope(!Z!, !X!, !Y!), :YOUR_TARGET_SRID)
```
Or alternatively, the `ST_TileEnvelope()` function supports an optional bounds

```sql
ST_TileEnvelope(!Z!, !X!, !Y!, :YOUR_BOUNDS)
```

[The documentation](https://postgis.net/docs/ST_TileEnvelope.html) defines the bounds as:

> The optional bounds parameter can be used to generate tiles in any coordinate system. It is a geometry that has the SRID and extent of the "Zoom Level zero" square within which the XYZ tile system is inscribed.


You mentioned using this custom function when calling `ST_AsMVTGeom`:

>  changed sql in tegola configuration to use this function instead of !BBOX! both when calling ST_AsMVTGeom as well as in WHERE clause.

You should be able to use the same `ST_TileEnvelope()` strategy above to pass in the geom to `ST_AsMVTGeom`. This would allow you to use only native PostGIS functions to achieve your desired outcome around querying and tile forming. Could you give this a try?

**Comment by tomass** [comment]
@ARolek thank you for information, here are my further thoughts.

> Ideally the `geom` package would be able to support different projections
> like you're exploring here. I'm not familiar enough with projection math to
> make these enhancements. Is this in your wheelhouse? If so, it would be
> a fantastic addition.

  There is no need to do any projection math here ;-) All heavylifting is done by PostGIS MVT functions.
  We only need to calculate BBOX correctly.
  If we could come up with a good/acceptable way of adding all required parts into configuration, I might be able to do the PR with some guidance.
  It would be simple, if it would be possible to reuse the SRIS configuration option like this:
  `SRID=###,[scale0resolution, scale1resolution, ...]`
  for example:
  `SRID=3346,[793.7515875,529.167725,264.583862,132.291931]`
  This would be all what has to be done in configuration to get custom projection tiles.
  But SRID is already used somehow, Tegola is currently trying to transform data from specified SRID to 4326/3857, right?

>> !BBOX! is mandatory, so I've added a pointless "and !BBOX! is not null" in WHERE clause

> Hmm, that's annoying. We should check for the use of `!BBOX!` or `!Z!` `!X!` `!Y!`

  If Tegola would support other projections this validation should stay as it is very useful, sql without the BBOX might cause some serious overwork.

> Did you have to implement a custom `calculate_my_projection_bbox(!z!,!x!,!y!)`
> function? I figured you could accomplish this 2 ways:
> ...

  _ST_TileEnvelope_ is very close to what is required, but at least in my case difference is not only in projection, but in scales as well. Mercator scales are calculated mathematically while in my case scales come from some old "tradition" and n+1 scale is not simple multiplication of previous one by 2. So without being able to specify all scales (resolutions for all scales) it is impossible to calculate the BBOX/Envelope simply by having zxy.

  I wonder how is it in @callsumzg case? Do you have some custom scales as well? Or are scale resolutions the same as for Mercator?

  Real scale resolutions:
  793.7515875031751,
  529.1677250021168, <--- diff from previous one 0,66
  264.5838625010584, <-- diff from previous one 0,5
  132.2919312505292, <-- diff from previous one 0,5
  66.1459656252646, <-- diff from previous one 0,5
  26.458386250105836, <-- diff from previous one 0,39

  This allows to have scales which resemble "old school" ones like 1:5000, 1:10000, 1:25000 (note the jump here is not x2), 1:50000 etc.

  Specifying scale resolutions is not something strange or new, it is a standard thing in say WMTS GetCapabilities.

**Comment by tomass** [comment]
Just for the record, if somebody has an answer or somebody later gets the same question.

I've thought about a variant where mercator scales would be used and thus standard st_tileenvelope could be used. Unfortunately the way st_tileenvelope works is you provide some square which will be your 0 scale tile. Then 1 scale would be 0 scale divided to four equal tiles and the same principal further.

This is completely different from what is done in OpenLayers where we have tile size - say 512 and when resolution - size of one pixel. If resolution is 0.5 when tile size is 256 meters x 256 meters. So our first tile is from 0,0 to 256,256 meters.

Therefore I do not currently see a nice way to "convert" from resolution way of calculating tiles (which is used in OL) to simple division used by PostGIS. In my understanding most projections would be "impossible" on mercator zoom 0. But I might be wrong here.

**Comment by ARolek** [comment]
@tomass sorry for the slow response here! The holidays + being a dad took over my free time ;-) 

> ST_TileEnvelope is very close to what is required, but at least in my case difference is not only in projection, but in scales as well. Mercator scales are calculated mathematically while in my case scales come from some old "tradition" and n+1 scale is not simple multiplication of previous one by 2. So without being able to specify all scales (resolutions for all scales) it is impossible to calculate the BBOX/Envelope simply by having zxy.

This is a key detail! Thanks for writing that out.

> If we could come up with a good/acceptable way of adding all required parts into configuration,

I like your proposal for a SRID -> Resolution array. I think this could work. I need to think about how we would want to implement this. This would essentially need to be an "override" for the native bbox transformations. 

> But SRID is already used somehow, Tegola is currently trying to transform data from specified SRID to 4326/3857, right?

Not sure I follow. Tegola only understand 3857 and 4326. 

Can you share the contents of your SQL function you referred to as `calculate_my_projection_bbox(!z!,!x!,!y!)`?

**Comment by tomass** [comment]
>> But SRID is already used somehow, Tegola is currently trying to transform data from specified SRID to 4326/3857, right?
> 
> Not sure I follow. Tegola only understand 3857 and 4326.

  Well, I remember I've tried simply specifying 3346 for the srid. Tegola complained that it does not know how to convert data from 3346 to 3857. Or something like that.

> Can you share the contents of your SQL function you referred to as `calculate_my_projection_bbox(!z!,!x!,!y!)`?

  Yes, I've placed all parts above, but for simplicity here is everything in one place. It's simply converting pixels to meters (pixels*resolution) and then displacing it to the right and up from some origin point (if tiles do not begin at position 0,0).

```
create or replace function tile_bbox(z int, x int, y int) returns geometry
language plpgsql as $$
declare
res float[] = ARRAY[ -- this is an array of resolutions for custom projection
  793.7515875031751,
  529.1677250021168,
  264.5838625010584,
  132.2919312505292,
  66.1459656252646,
  26.458386250105836,
  13.229193125052918,
  6.614596562526459,
  2.6458386250105836,
  1.3229193125052918
];
  c_tilesize int = 512;
  c_origin_x float = 0;
  c_origin_y float = 6500000;

l_tile_geom geometry;
begin
  l_tile_geom = st_setsrid(st_makepolygon(st_geomfromtext('LINESTRING(' ||
    ((x    ) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y    ) * res[z] * c_tilesize) || ',' ||
    ((x + 1) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y    ) * res[z] * c_tilesize) || ',' ||
    ((x + 1) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y + 1) * res[z] * c_tilesize) || ',' ||
    ((x    ) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y + 1) * res[z] * c_tilesize) || ',' ||
    ((x    ) * res[z] * c_tilesize) + c_origin_x || ' ' || c_origin_y - ((y    ) * res[z] * c_tilesize) ||
    ')')), 3346);

  return l_tile_geom;
end$$;
```

You then have the same numbers on OpenLayers side:
```
var lks_resolutions = [
  793.7515875031751,
  529.1677250021168,
  264.5838625010584,
  132.2919312505292,
  66.1459656252646,
  26.458386250105836,
  13.229193125052918,
  6.614596562526459,
  2.6458386250105836,
  1.3229193125052918];
```

And define a tileGrid like this with the same origin numbers:
```
var vectorGrid = new ol.tilegrid.TileGrid({
  extent: lks_extent, // [-3868431.3448, 3787209.7969000004, 3227425.9922, 9284025.053]
  resolutions: lks_resolutions,
  tileSize: [512,512],
  origin: [0,6500000]
});
```

And then you can create a vector tile source:
```
source: new ol.source.VectorTile({
  format: new ol.format.MVT(),
  tileGrid: vectorGrid,
  url: 'http://server:9090/maps/my_map/{z}/{x}/{y}',
  projection: 'EPSG:3346'
})
```

**Comment by callsumzg** [comment]
I finally found time to try out the custom Bbox function that @tomass developed in PostgreSQL 
Based on the grids we use for our WMTS in MapProxy, I tested two variants (or grids). One of them is based on the official standard of the Swiss government (**eCH-0056 norm**), see the official paper in German, pages **16/17** ([eCH-0056 PDF](https://www.ech.ch/sites/default/files/imce/eCH-Dossier/eCH-Dossier_PDF_Publikationen/Hauptdokument/STAN_d_DRA_2023-11-14_eCH-0056_V4.0.0_Anwendungsprofil%20Geodienste.pdf)), while the other one is a **legacy grid** that I inherited from my predecessor and, unfortunately, never questioned. I have no idea why it exists. Below are the two grid configurations:

### **Resolutions from official eCH-0056 standard for EPSG 2056**
```
4000, 2000, 1000, 500, 250, 100, 50, 20, 10,  
5, 2.5, 1, 0.5, 0.25, 0.1, 0.05
```
### "Legacy" resolutions, used for WMTS in MapProxy
```
26.45833332, 19.84374999, 13.22916666, 10.583333328,  
7.937499996, 6.61458333, 5.291666664, 3.968749998,  
3.307291665, 2.6458333322, 1.9843749999, 1.3229166666,  
0.7937499996, 0.6614583333, 0.5291666667, 0.3968749998,  
0.2645833333, 0.1984374999, 0.1322916667, 0.0793749999,  
0.0661458333, 0.0529166666
```

The official eCH-0056 grid worked perfectly, whereas the legacy grid only worked partially.

However, for both grids, I encountered a shift in the display. With the legacy grid, there was only a "small" shift (see screenshot), whereas for the eCH-0056 grid, the shift was much larger, with the tiles appearing somewhere on the moon.

### Before correction in Postgresql function (resolutions[z])

![Image](https://github.com/user-attachments/assets/07e433f7-7b6f-4441-a8e5-c043079ad72c)

### After correction in Postgresql function (resolutions[z+1])

![Image](https://github.com/user-attachments/assets/0f94d5c9-dd54-4555-8ab7-028a6a14a366)

After extensive troubleshooting and experimentation, I was able to fix the issue by modifying the PostgreSQL function, replacing the parameter "z" with "z+1". Apparently, the reason was that PostgreSQL uses a 1-based index. Because of this, the resolutions were incorrectly applied, leading to misaligned tile boundaries and distorted maps. After adjusting the PostgreSQL function accordingly, the display worked correctly. @tomass, can you explain this, and did you not encounter this problem?

Here is my adjusted PostgreSQL function using the eCH-0056 resolutions. Notice the "z+1" correction:

```sql
CREATE OR REPLACE FUNCTION tile_bbox_2056(z int, x int, y int)
RETURNS geometry AS
$$
DECLARE
  resolutions float[] := array[
    4000, 2000, 1000, 500, 250, 100, 50, 20, 10, 
    5, 2.5, 1, 0.5, 0.25, 0.1, 0.05
  ];
  
  tile_size int := 256;
  origin_x float := 2419995.75;
  origin_y float := 1350004.29;
  bbox geometry; 
BEGIN
  bbox := ST_SetSRID(
    ST_MakeEnvelope(
      origin_x + x * (resolutions[z+1] * tile_size),  -- Note z+1 for array index
      origin_y - (y + 1) * (resolutions[z+1] * tile_size),
      origin_x + (x + 1) * (resolutions[z+1] * tile_size),
      origin_y - y * (resolutions[z+1] * tile_size)
    ),
    2056
  );

  -- Debug output on postgres database stdout
  RAISE NOTICE 'Z:% X:% Y:% -> BBOX: %', z, x, y, ST_AsText(bbox);
  
  RETURN bbox;
END;
$$ LANGUAGE plpgsql IMMUTABLE;
```

For the legacy grid, I used the following PostgreSQL function:
```sql
CREATE OR REPLACE FUNCTION tile_bbox_2056(z int, x int, y int)
RETURNS geometry AS
$$
DECLARE
  resolutions float[] := array[
    26.45833332, 19.84374999, 13.22916666, 10.583333328, 
    7.937499996, 6.61458333, 5.291666664, 3.968749998,
    3.307291665, 2.6458333322, 1.9843749999, 1.3229166666,
    0.7937499996, 0.6614583333, 0.5291666667, 0.3968749998,
    0.2645833333, 0.1984374999, 0.1322916667, 0.0793749999,
    0.0661458333, 0.0529166666
  ];
  
  tile_size int := 512;
  origin_x float := 2670000;
  origin_y float := 1235000;
  bbox geometry;  
BEGIN
 -- same code as above
END;
$$ LANGUAGE plpgsql IMMUTABLE;
```

In Openlayers, i just updated the map resolutions, extent, origin and tilesize accordingly, just as @tomass did in his example.

### Issue with the Legacy Grid
When using this legacy grid, I encountered an issue I couldn't resolve. Tegola returns an "invalid X/Y value" error when requesting the following tiles:

```
0/0/1  
0/1/0  
0/1/1  
0/2/0  
0/2/1  
1/2/1  
1/2/0  
1/0/2  
1/2/2  
1/1/2  
1/2/0  
```

All other tiles work fine, for example:

```
0/0/0  
1/1/1  
1/0/1  
1/1/0  
1/0/0  
2/2/2  
2/2/3  
2/1/3  
2/1/2  
2/3/3  
2/3/2  
2/0/3  
2/0/2  
...
```

The error Tegola throws is handled in handle_map_layer_zxy.go, specifically at line [74](https://github.com/go-spatial/tegola/blob/ae36ce929ebf73655e9a142707805231a9323966/server/handle_map_layer_zxy.go#L74) and [84](https://github.com/go-spatial/tegola/blob/ae36ce929ebf73655e9a142707805231a9323966/server/handle_map_layer_zxy.go#L84).

### Understanding the Issue with my legacy WMTS grid
Tegola's internal tile validation enforces the standard slippy map grid (X/Y ≤ 2ᶻ-1), but my custom grid requires different limits. This conflict happens because:

1. Standard Grid Assumption
Tegola's code uses math.Exp2(z) - 1 to calculate the maximum X/Y values:
Example: At Z=0, the max X/Y=0, but my custom grid allows X=2, Y=1

2. Custom Grid Reality of my legacy grid
My legacy scale allows larger X/Y values due to one or a combination of the following reason:
- special Non-power-of-2 resolutions
- Specific geographic extent (2670000,1212000 → 2698000,1235000)
- 512px tile size

**However, if this custom legacy scale doesn't work, I really don't mind replacing it. 
According to the official eCH-0056 standard, we shouldn't be using it anyway for our WMTS (and neither Vector Tiles)**

@ARolek and @iwpnd: How much effort would it take to implement custom grids in Tegola?

**Comment by tomass** [comment]
Hello, sorry for late response. Shift in zoom level might be because your tiles are 256 and not 512, this is a usual thing, extent of 512 tile is the same as 256 of a previous zoom level. For example if you want to switch from a 256px size layer to 512px size layer so that they display "the same" place, you will have to subtract 1 from the zoom level value (it is the math of extent calculation so there is no difference here if we talk about raster or vtiles layer).

Shift in legacy was "small" because resolutions are very close (not x2 or close to that as for current official resolution set).

But if I understand correctly, Tegola only supports 512px tile size, as that is kind of a standard for vector tiles? (Or you would have to do the zoom shift solution you've already found on the front size).

**Comment by tomass** [comment]
@ARolek what do you think about actually implementing this? I could try to prepare a series of small PR's advancing one step at a time. We just need a decision on where are tile boundaries (of non 3857) calculated: tegola or database:
* if database, allow not providing BBOX in sql
* if tegola, we need to decide where and how projection information is entered, we need to enter: origin [x,y], scale resolutions [...]

(and this would only work for postgis mvt provider...)

**Comment by ARolek** [comment]
@tomass apologies for the slow response. I'm open to this feature being added. I believe @NivGreenstein has been [working on a fork of tegola](https://github.com/go-spatial/tegola/issues/1104#issuecomment-4671876524) to support this functionality so they might be interested in helping with the effort.

> We just need a decision on where are tile boundaries (of non 3857) calculated

It seems to me that pushing the calculation to the database might be the easiest first step. Which direction are you leaning?

> prepare a series of small PR's advancing one step at a time

If you have a chance, I would be interested in the PR boundaries you're thinking about.

**Comment by NivGreenstein** [comment]
> [@tomass](https://github.com/tomass) apologies for the slow response. I'm open to this feature being added. I believe [@NivGreenstein](https://github.com/NivGreenstein) has been [working on a fork of tegola](https://github.com/go-spatial/tegola/issues/1104#issuecomment-4671876524) to support this functionality so they might be interested in helping with the effort.
> 
> > We just need a decision on where are tile boundaries (of non 3857) calculated
> 
> It seems to me that pushing the calculation to the database might be the easiest first step. Which direction are you leaning?
> 
> > prepare a series of small PR's advancing one step at a time
> 
> If you have a chance, I would be interested in the PR boundaries you're thinking about.

Hey @ARolek, thanks for involving me in this discussion!

Yes, I’ve been working on a [fork of Tegola that supports EPSG:4326 WorldCRS84Quad tile requests](https://github.com/NivGreenstein/tegola/pull/2).

My current implementation introduces a `tile_srid` option under `[[maps]]`. Each Tegola map is associated with a single tile grid, so its `/{z}/{x}/{y}` endpoint interprets the coordinates according to that grid.

Currently:

* `3857` remains the default and uses WebMercatorQuad.
* `4326` uses WorldCRS84Quad.
* A single map cannot currently expose multiple tile matrix sets.
* In my POCs, I haven’t observed a meaningful performance difference compared with Tegola’s default behavior.

Regarding where tile boundaries should be calculated: my fork currently calculates them in Tegola. The tile grid affects more than the database query - it is also needed for validating `z/x/y`, calculating buffered extents, seeding tiles, and handling cache paths consistently.

That said, I agree that pushing the calculation into PostGIS could be the smallest useful first step, especially if the initial scope is limited to the `mvt_postgis` provider. It would let us add practical support incrementally without requiring the entire tile-grid architecture to be generalized in the first PR.

For the longer-term design, I think Tegola should understand the tile matrix set being served, while the database remains responsible for querying the data and encoding the MVT. MVT itself is not inherently tied to EPSG:3857; that is primarily a convention adopted by much of the existing ecosystem.

As a longer-term goal, I would also love to see Tegola support [OGC API – Tiles](https://ogcapi.ogc.org/tiles/). Making the `tileMatrixSetId` explicit would provide a standardized way to discover and request tiles in different projections, rather than relying on an implicit interpretation of `/{z}/{x}/{y}`.

[developmentseed/tipg](https://github.com/developmentseed/tipg) is a useful reference for this behavior.

Tegola is a great project, and I’d be very happy to actively help design and implement this.
